game-core/src/main/java/games/strategy/triplea/delegate/battle/steps/retreat/OffensiveGeneralRetreat.java

```diff
@@ -0,0 +1,129 @@
+package games.strategy.triplea.delegate.battle.steps.retreat;
+
+import static games.strategy.triplea.delegate.battle.BattleStepStrings.ATTACKER_WITHDRAW;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.Territory;
+import games.strategy.engine.data.Unit;
+import games.strategy.engine.delegate.IDelegateBridge;
+import games.strategy.triplea.Properties;
+import games.strategy.triplea.delegate.ExecutionStack;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.battle.BattleActions;
+import games.strategy.triplea.delegate.battle.BattleState;
+import games.strategy.triplea.delegate.battle.MustFightBattle.RetreatType;
+import games.strategy.triplea.delegate.battle.steps.BattleStep;
+import games.strategy.triplea.delegate.battle.steps.RetreatChecks;
+import java.util.Collection;
+import java.util.List;
+import java.util.Set;
+import javax.annotation.Nullable;
+import lombok.AllArgsConstructor;
+import lombok.Value;
+import org.triplea.java.collections.CollectionUtils;
+
+@AllArgsConstructor
+public class OffensiveGeneralRetreat implements BattleStep {
+
+  final BattleState battleState;
+
+  final BattleActions battleActions;
+
+  @Override
+  public List<String> getNames() {
+    if (canAttackerRetreat()
+        || canAttackerRetreatSeaPlanes()
+        || (battleState.isAmphibious()
+            && (canAttackerRetreatPartialAmphib() || canAttackerRetreatAmphibPlanes()))) {
+      return List.of(battleState.getAttacker().getName() + ATTACKER_WITHDRAW);
+    }
+    return List.of();
+  }
+
+  @Override
+  public Order getOrder() {
+    return Order.OFFENSIVE_GENERAL_RETREAT;
+  }
+
+  @Override
+  public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+    retreatUnits(bridge);
+  }
+
+  public void retreatUnits(final IDelegateBridge bridge) {
+    if (battleState.isOver()) {
```

Comment:
Does this condition happen often? TBH, it seems a bit like a bug. If the battle is over, the caller perhaps did not realize that and should not have made this call to begin with.

Reply:
Yes.  How the stack work is that at the beginning of the round, all of the IExecutables/BattleSteps are created and added to the stack.  Then, it goes through each of the items in the step.  There isn't currently a way to short circuit the stack when the battle is over.  So, this step already is in the stack and will execute even if the battle finished in a previous step.

Reply:
Thanks for the context. I don't find that terribly satisfying, but I would accept that wiring a 'Predicate' for battle is over to the execution stack at this juncture is potentially not easy and out of scope of this update.